packages/dialect-react-sdk-blockchain-solana/context.tsx

```diff
@@ -0,0 +1,125 @@
+import {
+  DialectSolanaWalletAdapter,
+  Solana,
+  SolanaSdkFactory,
+} from '@dialectlabs/blockchain-sdk-solana';
+import {
+  DialectContextProvider,
+  DialectContextProviderProps,
+  DialectWalletStatesHolder,
+} from '@dialectlabs/react-sdk';
+import { PublicKey } from '@solana/web3.js';
+import { useCallback, useEffect, useMemo } from 'react';
+import type { SolanaConfigProps } from './types';
+
+interface DialectSolanaSdkProps
+  extends Omit<DialectContextProviderProps<Solana>, 'blockchainSdkFactory'> {
+  solanaConfig: SolanaConfigProps;
+}
+
+const SolanaBlockchainSdkWrapper = ({
+  solanaConfig,
+  ...props
+}: DialectSolanaSdkProps) => {
+  const {
+    walletConnected: { set: setWalletConnected },
+    connectionInitiatedState: { set: setConnectionInitiated },
+    isSigningFreeTransactionState: { set: setIsSigningFreeTransaction },
+    isSigningMessageState: { set: setIsSigningMessage },
+    hardwareWalletForcedState: { get: isHardwareWalletForced },
+    isEncryptingState: { set: setIsEncrypting },
+  } = DialectWalletStatesHolder.useContainer();
+
+  const wrapDialectWallet = useCallback(
+    (adapter: DialectSolanaWalletAdapter): DialectSolanaWalletAdapter => {
+      return {
+        ...adapter,
+        signTransaction: adapter.signTransaction
+          ? async (tx) => {
+              const isFreeTx =
+                tx.recentBlockhash === PublicKey.default.toString();
+              if (isFreeTx) {
+                setIsSigningFreeTransaction(true);
+              }
+              try {
+                return await adapter.signTransaction!(tx);
+              } catch (e) {
+                if (isFreeTx) {
+                  // assuming free tx is the tx for auth
+                  setConnectionInitiated(false);
+                }
+                throw e;
+              } finally {
+                if (isFreeTx) {
+                  setIsSigningFreeTransaction(false);
+                }
+              }
+            }
+          : undefined,
+        signMessage:
+          !isHardwareWalletForced && adapter.signMessage
+            ? async (msg) => {
+                setIsSigningMessage(true);
+                try {
+                  return await adapter.signMessage!(msg);
+                } catch (e) {
+                  // assumint sign message used only for auth
+                  setConnectionInitiated(false);
+                  throw e;
+                } finally {
+                  setIsSigningMessage(false);
+                }
+              }
+            : undefined,
+        diffieHellman: adapter.diffieHellman
+          ? async (...args) => {
+              setIsEncrypting(true);
+              try {
+                return await adapter.diffieHellman!(...args);
```

Comment:
yeah, I know, but for some reason it doesn't allow to

<img width="596" alt="Screenshot 2022-10-14 at 18 27 26" src="https://user-images.githubusercontent.com/5163918/195884053-84e1e9f0-0691-484f-9054-9ea97481b99a.png">

Reply:
ohhhh, i know why. Since the function is async, at the time of the promise execution, `adapter.diffieHellman` technically can be undefined, since it's used by reference.

Reply:
lol, i died